channeldb/graph_cache.go

```diff
@@ -0,0 +1,331 @@
+	toNode route.Vertex, edge1 bool) {
+
+	// If a policy's node is nil, we can't cache it yet as that would lead
+	// to problems in pathfinding.
+	if policy.Node == nil {
+		// TODO(guggero): Fix this problem!
+		log.Warnf("Cannot cache policy because of missing node (from "+
+			"%x to %x)", fromNode[:], toNode[:])
+		return
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	updatePolicy := func(channels []*DirectedChannel) {
+		for idx, channel := range channels {
+			// Is this the channel in question?
+			if channel.ChannelID == policy.ChannelID {
+				// Edge 1 is defined as the policy for the
+				// direction of node1 to node2.
+				switch {
+				// This is node 1, and it is edge 1, so this is
+				// the outgoing policy for node 1.
+				case channel.IsNode1 && edge1:
+					channels[idx].OutPolicy = policy
+
+				// This is node 2, and it is edge 2, so this is
+				// the outgoing policy for node 2.
+				case !channel.IsNode1 && !edge1:
+					channels[idx].OutPolicy = policy
+
+				// The other two cases left mean it's the
+				// inbound policy for the node.
+				default:
+					channels[idx].InPolicy = policy
+				}
+				return
+			}
+		}
+	}
+
+	updatePolicy(c.nodeChannels[fromNode])
+	updatePolicy(c.nodeChannels[toNode])
+}
+
+// RemoveNode completely removes a node and all its channels (including the
+// peer's side).
+func (c *GraphCache) RemoveNode(node route.Vertex) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	delete(c.nodeFeatures, node)
+
+	// First remove all channels from the other nodes' lists.
+	for _, channel := range c.nodeChannels[node] {
+		c.removeChannelIfFound(channel.OtherNode, channel.ChannelID)
+	}
+
+	// Then remove our whole node completely.
+	delete(c.nodeChannels, node)
+}
+
+// RemoveChannel removes a single channel between two nodes.
+func (c *GraphCache) RemoveChannel(node1, node2 route.Vertex, chanID uint64) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// Remove that one channel from both sides.
+	c.removeChannelIfFound(node1, chanID)
+	c.removeChannelIfFound(node2, chanID)
+}
+
+// removeChannelIfFound removes a single channel from one side.
+func (c *GraphCache) removeChannelIfFound(node route.Vertex, chanID uint64) {
+	for i := 0; i < len(c.nodeChannels[node]); i++ {
+		if c.nodeChannels[node][i].ChannelID == chanID {
+			// We found the channel. Shift everything to the left
+			// by one then return since we only expect to find a
+			// single channel.
+			c.nodeChannels[node] = append(
+				c.nodeChannels[node][:i],
+				c.nodeChannels[node][i+1:]...,
+			)
+			return
+		}
+	}
+}
+
+// UpdateChannel updates the channel edge information for a specific edge.
+func (c *GraphCache) UpdateChannel(info *ChannelEdgeInfo) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	for idx := range c.nodeChannels[info.NodeKey1Bytes] {
+		channel := c.nodeChannels[info.NodeKey1Bytes][idx]
+		if channel.ChannelID == info.ChannelID {
+			channel.Capacity = info.Capacity
+			channel.OtherNode = info.NodeKey2Bytes
```

Comment:
The `ChannelEdgeInfo` doesn't contain any policies. This whole method probably is a no-op most of the time. But wanted to keep it in case we ever update the capacity or peer node key (which sounds unlikely, but still).